### DISCUSSION TITLE
Flexibly Create Nested Database Entries from Incoming Pydantic/SQLModels

### DISCUSSION THREAD

**Discussion by hay-kot** [fastapi/sqlmodel#1918] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Sudo Code Based on Examples in Docs

class Team(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    headquarters: str


    heroes: List["Hero"] = Relationship(back_populates="team")



class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None

    team_id: Optional[int] = Field(default=None, foreign_key="team.id")

    team: Optional[Team] = Relationship(back_populates="heroes")


payload = {
    "name": "Team Name",
    "headquarters": "Whereever".
    "heroes": [
        "name": "Name 1"
        // Other Requied Fields... 👇
    ]
}

with Session(engine) as session:
    Team.create_all_nested(session, payload) # or something?
```

### Description

I would like to do what is described in [FastAPI issue #2194](https://github.com/tiangolo/fastapi/issues/2194)

> How to make nested sqlalchemy models from nested pydantic models (or python dicts) in a generic way and write them to the database in "one shot".

In the example above, I'd like to pass in the payload to a method and the following to occur.
* Create new Team entry
* Create Hero entry and/or Relate the existing Hero to the Team

Similarly, I'd like the same to happen on update. Effectively making writing to the SQL database akin to writing to MongoDB

I don't believe this is supported or haven't gotten it to work, but my main questions are.
1. Is this supported? 
2. If no, is this a use-case you've thought of? 
3. Are you interested in a PR to support this either as a utility method or some sort of decorator? 

Loving working with this so far, thanks for all your hard work! 

### Operating System

macOS

### SQLModel Version

0.0.3

### Python Version

3.9.6

### Additional Context

I have accomplished this with SQLAlchemy in the past by using an `auto_init` decarator.

```python
from functools import wraps
from typing import Union

from sqlalchemy.orm import MANYTOMANY, MANYTOONE, ONETOMANY


def handle_one_to_many_list(relation_cls, all_elements: list[dict]):
    elems_to_create = []
    updated_elems = []

    for elem in all_elements:
        elem_id = elem.get("id", None)

        existing_elem = relation_cls.get_ref(match_value=elem_id)

        if existing_elem is None:

            elems_to_create.append(elem)

        else:
            for key, value in elem.items():
                setattr(existing_elem, key, value)

            updated_elems.append(existing_elem)

    new_elems = []
    for elem in elems_to_create:
        new_elems = [relation_cls(**elem) for elem in all_elements]

    return new_elems


def auto_init(exclude: Union[set, list] = None):  # sourcery no-metrics
    """Wraps the `__init__` method of a class to automatically set the common
    attributes.

    Args:
        exclude (Union[set, list], optional): [description]. Defaults to None.
    """

    exclude = exclude or set()
    exclude.add("id")

    def decorator(init):
        @wraps(init)
        def wrapper(self, *args, **kwargs):  # sourcery no-metrics
            """
            Custom initializer that allows nested children initialization.
            Only keys that are present as instance's class attributes are allowed.
            These could be, for example, any mapped columns or relationships.

            Code inspired from GitHub.
            Ref: https://github.com/tiangolo/fastapi/issues/2194
            """
            cls = self.__class__
            model_columns = self.__mapper__.columns
            relationships = self.__mapper__.relationships

            session = kwargs.get("session", None)

            for key, val in kwargs.items():
                if key in exclude:
                    continue

                if not hasattr(cls, key):
                    continue
                    # raise TypeError(f"Invalid keyword argument: {key}")

                if key in model_columns:
                    setattr(self, key, val)
                    continue

                if key in relationships:
                    relation_dir = relationships[key].direction.name
                    relation_cls = relationships[key].mapper.entity
                    use_list = relationships[key].uselist

                    if relation_dir == ONETOMANY.name and use_list:
                        instances = handle_one_to_many_list(relation_cls, val)
                        setattr(self, key, instances)

                    if relation_dir == ONETOMANY.name and not use_list:
                        instance = relation_cls(**val)
                        setattr(self, key, instance)

                    elif relation_dir == MANYTOONE.name and not use_list:
                        if isinstance(val, dict):
                            val = val.get("id")

                            if val is None:
                                raise ValueError(f"Expected 'id' to be provided for {key}")

                        if isinstance(val, (str, int)):
                            instance = relation_cls.get_ref(match_value=val, session=session)
                            setattr(self, key, instance)

                    elif relation_dir == MANYTOMANY.name:

                        if not isinstance(val, list):
                            raise ValueError(f"Expected many to many input to be of type list for {key}")

                        if len(val) > 0 and isinstance(val[0], dict):
                            val = [elem.get("id") for elem in val]

                        instances = [relation_cls.get_ref(elem, session=session) for elem in val]
                        setattr(self, key, instances)

            return init(self, *args, **kwargs)

        return wrapper

    return decorator

```

## Usage

```python

class AdminModel(SqlAlchemyBase, BaseMixins):
    name = Column(String, index=True)
    email = Column(String, unique=True, index=True)
    password = Column(String)
    is_superuser = Column(Boolean(), default=False)

    @auto_init(exclude={'is_superuser'})
    def __init__(self, **_):
        this.is_superuser = false

    @classmethod
    def get_ref(cls, match_value: str, match_attr: str = "id"):
        with SessionLocal() as session:
            eff_ref = getattr(cls, match_attr)
            return session.query(cls).filter(eff_ref == match_value).one_or_none()

```decorator

**Comment by khialb32** [comment]
Hello @hay-kot , i have tried your solution on nested models, and it works just fine, I will try with more nested pedantic models and see if it works, yet I would greatly appreciate if you could provide some explanation(documentation) of what you did because although I read it many times, I am still afraid to use it for production.

**Comment by hobbsAU** [comment]
SQLModel is fantastic as it cuts so much time out managing pydantic schemas and sqlalchemy models.

I too would like to know how to create database entries with incoming nested SQLModels.

Does anyone know how to do this yet?

**Comment by woprandi** [comment]
It's even worse than use separate pydantic and sqlalchemy models because the nested list on the payload received from client are always empty. So it's not even possible to create each element separarely

**Comment by scd75** [comment]
Agree with @hay-kot : it would be great to have the ability to instantiate nested objects from a nested SQLModel instance.
in the same vein as @hay-kot : i used with the past a custom constructor for SQLAlchemy Base class, and also proposed that to SQLAlchemey repo:
https://github.com/tiangolo/fastapi/issues/2194
https://github.com/sqlalchemy/sqlalchemy/discussions/6322
@tiangolo : any idea on how this could be achieved? Thanks!

**Comment by Fanna1119** [comment]
This would such an amazing feature. Any update on this?

**Comment by Sancho66** [comment]
Any update on this ?

**Comment by Sillocan** [comment]
This is a major feature that is missing from sqlmodel that is preventing myself from migrating a larger project from marshmallow. There is a lot of boilerplate code that needs to be added to handle nested objects

**Comment by zrajna** [comment]
I would also love to see this addition to the awesome SQLModel library.

**Comment by 50Bytes-dev** [comment]
Started working with this wonderful library and encountered the same problem. Any updates?

**Comment by micoloth** [comment]
+1 on this. 

First of all, this library looks amazing!

Last year, I was using Sqlalchemy and Pydantic models separately, but this library would be such a quality of life improvement..

Unfortunately, the fact that this feature is missing is a deal breaker, so i cannot start using it.







Last year, I also built a custom recursive solution to do this in Sqlalchemy+Pydantic, just like OP did.

But, I don't think it would be easy/a good idea to port it to SqlModels...

Any updates would be greatly appreciated!

**Comment by s-froghyar** [comment]
Any update on this? Would love to use `hydra` as a configuration management the way it is meant to be used (hierarchical and out of the box configurable with separation of concerns) rather than making a single flat yaml for all configs

**Comment by whugarra** [comment]
There any temporary guidance from best practices to avoid this issue?

**Comment by dataengineeringatfunderzgroup** [comment]
any updates on it? this is something that will be amazing for SQLModel to have.

**Comment by elliswilcox** [comment]
Hi, is there any updates or fixes for this?
Still having this issue in Nov 2024, a good 3 years after this issue was originally created.
I'm sure a lot of people (including me) would benefit from this being fixed.

@tiangolo

**Comment by UmarBelloKanwa** [comment]
Still facing these issues is there any updates about?

**Comment by UmarBelloKanwa** [comment]
I successfully solved all these problems in one line. Just by converting my pydantic class models to dictionary and storing the dictionary in the database through SQLModel"s, class table=True

**Comment by tutebatti** [comment]
@UmarBelloKanwa Can you elaborate on this?

**Comment by elliswilcox** [comment]
> I successfully solved all these problems in one line. Just by converting my pydantic class models to dictionary and storing the dictionary in the database through SQLModel"s, class table=True

Could you provide a short code sample for this? I'm not 100% sure what you mean.

**Comment by UmarBelloKanwa** [comment]
> > I successfully solved all these problems in one line. Just by converting my pydantic class models to dictionary and storing the dictionary in the database through SQLModel"s, class table=True
> 
> Could you provide a short code sample for this? I'm not 100% sure what you mean.

You have to convert your base model to 
dictionary using model_dump() - pydantic.

Or look to this tutorial:
https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/#columns-and-inheritance-with-multiple-models

**Comment by Guddi11** [comment]
Hi @UmarBelloKanwa,
I think you triggered some of this issues' followers quite a bit with your teaser. Could you provide a MVE of how you managed to get your nested schemas to be validated and correctly fed into the ORM? Thanks a bunch.

I am also stuck at this very problem and building quite a complex data structure I'd be delighted if SQLModel provided some way to simplify nested model deserialization.

Merry Xmas everyone.

**Comment by muddassir-lateef** [comment]
+1 @tiangolo any updates for this feature. This is a great library but would love to have this feature.

**Comment by UmarBelloKanwa** [comment]
> Hi [@UmarBelloKanwa](https://github.com/UmarBelloKanwa), I think you triggered some of this issues' followers quite a bit with your teaser. Could you provide a MVE of how you managed to get your nested schemas to be validated and correctly fed into the ORM? Thanks a bunch.
> 
> I am also stuck at this very problem and building quite a complex data structure I'd be delighted if SQLModel provided some way to simplify nested model [deserialization.]
> 
> Merry Xmas everyone.

Check models and schemas on [link](https://github.com/UmarBelloKanwa/Shopark-current/tree/main/api)

**Comment by Sillocan** [comment]
> > Hi [@UmarBelloKanwa](https://github.com/UmarBelloKanwa), I think you triggered some of this issues' followers quite a bit with your teaser. Could you provide a MVE of how you managed to get your nested schemas to be validated and correctly fed into the ORM? Thanks a bunch.
> > I am also stuck at this very problem and building quite a complex data structure I'd be delighted if SQLModel provided some way to simplify nested model [deserialization.]
> > Merry Xmas everyone.
> 
> Check models and schemas on [link](https://github.com/UmarBelloKanwa/Shopark-current/tree/main/api)

Link isn't public

**Comment by nathggns** [comment]
This makes it quite difficult to use SQLModel in LLM applications, as we need to build out our models from JSON output from an LLM. There doesn't seem to be an obviously supported way to do this right now.

**Comment by davletovalmir** [comment]
I ended up inheriting my `SQLModel` models from custom `BaseTable` class that is defined as shown below. Feel free to use and extend! It helped a lot in my project, hope someone else could also benefit from it!

```python
class BaseTable(SQLModel):
    __abstract__ = True

    id: Optional[int] = Field(default=None, primary_key=True)

    # your other default columns / fields / methods here

    @classmethod
    def from_json(cls, json_str: str) -> Self:
        return cls.from_dict(json.loads(json_str))

    @classmethod
    def from_dict(cls, data: dict) -> Self:
        values = {}

        for field_name, raw_value in data.items():
            if field_name in {"id", "created_at", "updated_at"}:
                continue

            field_value = cls._build_field(field_name, raw_value)
            if field_value is None and raw_value is not None:  # skipped value
                continue
            values[field_name] = field_value

        return cls(**values)

    @classmethod
    def _build_field(cls, field_name: str, raw_value: Any) -> Any:
        if field_name in cls.model_fields:
            return raw_value

        rel = cls.__mapper__.relationships.get(field_name)
        if not rel:
            return None

        is_belongs_to_rel = f"{field_name}_id" in cls.model_fields
        if is_belongs_to_rel:
            return None

        FieldModel = rel.mapper.class_

        if rel.uselist:
            sub_items = []
            for item in raw_value:
                if isinstance(item, dict):
                    sub_items.append(FieldModel.from_dict(item))
                elif isinstance(item, FieldModel):
                    sub_items.append(item)
                else:
                    continue
            return sub_items

        if isinstance(raw_value, dict):
            return FieldModel.from_dict(raw_value)
        if isinstance(raw_value, FieldModel):
            return raw_value

        return None
```

**Comment by gzu300** [comment]
This is related to [https://github.com/tiangolo/sqlmodel/discussions/1116](https://github.com/tiangolo/sqlmodel/discussions/1116), where nested creation using SQLModel ORM models (`table=True`) was discussed.

Inspired by that, I implemented a cleaner pattern using **Helper models** that transform nested `Create` schemas into fully ORM-compatible structures. The key idea is to use `@field_validator(..., mode="after")` in dedicated helper layers — keeping `Create` schemas, transformation logic, and `ORM` models cleanly separated.

✅ The final ORM object is still created via `UserModel.model_validate(...)`.

### ✅ Minimal Working Example

```python
from sqlmodel import SQLModel, Field, Relationship
from pydantic import field_validator

# --- Input Schemas ---
class AddressCreate(SQLModel):
    city: str

class UserCreate(SQLModel):
    name: str
    address: AddressCreate

# --- ORM Models ---
class AddressModel(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    city: str

class UserModel(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str
    address_id: int | None = Field(default=None, foreign_key="addressmodel.id")
    address: AddressModel = Relationship()

# --- Helper Models ---
class AddressCreateHelper(AddressCreate):
    @field_validator("city", mode="after")
    @classmethod
    def to_model(cls, v, values):
        return AddressModel(city=v)

class UserCreateHelper(UserCreate):
    @field_validator("address", mode="after")
    @classmethod
    def to_model(cls, v):
        return AddressModel.model_validate(AddressCreateHelper.model_validate(v))

# --- Usage ---
payload = {"name": "Alice", "address": {"city": "Berlin"}}

# Validate and transform through helper layer
transformed = UserCreateHelper.model_validate(payload)

# Final ORM construction
user_model = UserModel.model_validate(transformed)
```

✅ `UserModel.model_validate(...)` receives fully transformed nested ORM instances.

This pattern is scalable, safe, and keeps each layer focused. Let me know if you'd like a generalized recursive helper approach.

**Comment by WilliamDEdwards** [comment]
> This is related to [#1116](https://github.com/fastapi/sqlmodel/discussions/1116), where nested creation using SQLModel ORM models (`table=True`) was discussed.
> 
> Inspired by that, I implemented a cleaner pattern using **Helper models** that transform nested `Create` schemas into fully ORM-compatible structures. The key idea is to use `@field_validator(..., mode="after")` in dedicated helper layers — keeping `Create` schemas, transformation logic, and `ORM` models cleanly separated.
> 
> ✅ The final ORM object is still created via `UserModel.model_validate(...)`.
> ### ✅ Minimal Working Example
> 
> from sqlmodel import SQLModel, Field, Relationship
> from pydantic import field_validator
> 
> # --- Input Schemas ---
> class AddressCreate(SQLModel):
>     city: str
> 
> class UserCreate(SQLModel):
>     name: str
>     address: AddressCreate
> 
> # --- ORM Models ---
> class AddressModel(SQLModel, table=True):
>     id: int | None = Field(default=None, primary_key=True)
>     city: str
> 
> class UserModel(SQLModel, table=True):
>     id: int | None = Field(default=None, primary_key=True)
>     name: str
>     address_id: int | None = Field(default=None, foreign_key="addressmodel.id")
>     address: AddressModel = Relationship()
> 
> # --- Helper Models ---
> class AddressCreateHelper(AddressCreate):
>     @field_validator("city", mode="after")
>     @classmethod
>     def to_model(cls, v, values):
>         return AddressModel(city=v)
> 
> class UserCreateHelper(UserCreate):
>     @field_validator("address", mode="after")
>     @classmethod
>     def to_model(cls, v):
>         return AddressModel.model_validate(AddressCreateHelper.model_validate(v))
> 
> # --- Usage ---
> payload = {"name": "Alice", "address": {"city": "Berlin"}}
> 
> # Validate and transform through helper layer
> transformed = UserCreateHelper.model_validate(payload)
> 
> # Final ORM construction
> user_model = UserModel.model_validate(transformed)
> 
> ✅ `UserModel.model_validate(...)` receives fully transformed nested ORM instances.
> 
> This pattern is scalable, safe, and keeps each layer focused. Let me know if you'd like a generalized recursive helper approach.

Don't pollute the internet with LLM crap please. Thanks!

**Comment by gzu300** [comment]
> > This is related to [#1116](https://github.com/fastapi/sqlmodel/discussions/1116), where nested creation using SQLModel ORM models (`table=True`) was discussed.
> > Inspired by that, I implemented a cleaner pattern using **Helper models** that transform nested `Create` schemas into fully ORM-compatible structures. The key idea is to use `@field_validator(..., mode="after")` in dedicated helper layers — keeping `Create` schemas, transformation logic, and `ORM` models cleanly separated.
> > ✅ The final ORM object is still created via `UserModel.model_validate(...)`.
> > ### ✅ Minimal Working Example
> > from sqlmodel import SQLModel, Field, Relationship
> > from pydantic import field_validator
> > # --- Input Schemas ---
> > class AddressCreate(SQLModel):
> > city: str
> > class UserCreate(SQLModel):
> > name: str
> > address: AddressCreate
> > # --- ORM Models ---
> > class AddressModel(SQLModel, table=True):
> > id: int | None = Field(default=None, primary_key=True)
> > city: str
> > class UserModel(SQLModel, table=True):
> > id: int | None = Field(default=None, primary_key=True)
> > name: str
> > address_id: int | None = Field(default=None, foreign_key="addressmodel.id")
> > address: AddressModel = Relationship()
> > # --- Helper Models ---
> > class AddressCreateHelper(AddressCreate):
> > @field_validator("city", mode="after")
> > [@classmethod](https://github.com/classmethod)
> > def to_model(cls, v, values):
> > return AddressModel(city=v)
> > class UserCreateHelper(UserCreate):
> > @field_validator("address", mode="after")
> > [@classmethod](https://github.com/classmethod)
> > def to_model(cls, v):
> > return AddressModel.model_validate(AddressCreateHelper.model_validate(v))
> > # --- Usage ---
> > payload = {"name": "Alice", "address": {"city": "Berlin"}}
> > # Validate and transform through helper layer
> > transformed = UserCreateHelper.model_validate(payload)
> > # Final ORM construction
> > user_model = UserModel.model_validate(transformed)
> > ✅ `UserModel.model_validate(...)` receives fully transformed nested ORM instances.
> > This pattern is scalable, safe, and keeps each layer focused. Let me know if you'd like a generalized recursive helper approach.
> 
> Don't pollute the internet with LLM crap please. Thanks!

@WilliamDEdwards

1. I agree — this approach isn’t ideal, and it would definitely be better if the library itself provided native support for this feature. Still, I wanted to share one possible solution that works in the meantime.

2. For transparency: yes, I used an LLM to polish the wording of the previous comment (mainly for grammar and clarity), but the implementation and idea are entirely my own. I thought that’s exactly what these tools are meant for — improving communication, not generating the core logic.

3. And yes, again — I used an LLM to polish this comment too 🙂. In the end, we should focus on the issue itself, as long as we understand each other.
